File: sdk/go/configstore/clone.go
package configstore

import (
	"reflect"

	"github.com/Suhaibinator/kms/sdk/go/kmsclient"
)

var secretType = reflect.TypeFor[kmsclient.Secret]()

type cloneVisit struct {
	typ  reflect.Type
	kind reflect.Kind
	ptr  uintptr
	len  int
	cap  int
}

// Clone makes a recursive candidate-preparation copy of supported config
// values. In particular, kmsclient.Secret is copied through Secret.Clone so
// its plaintext backing buffer is never shared. Clone is intended for defaults,
// candidate construction, and defensive reports—not generated hot-path
// getters.
func Clone[T any](value T) T {
	cloned := cloneValue(reflect.ValueOf(value), make(map[cloneVisit]reflect.Value))
	if !cloned.IsValid() {
		var zero T
		return zero
	}
	return cloned.Interface().(T)
}

func cloneValue(value reflect.Value, seen map[cloneVisit]reflect.Value) reflect.Value {
	if !value.IsValid() {
		return reflect.Value{}
	}
	if value.Type() == secretType && value.CanInterface() {
		secret := value.Interface().(kmsclient.Secret).Clone()
		return reflect.ValueOf(secret)
	}

	switch value.Kind() {
	case reflect.Interface:
		if value.IsNil() {
			return reflect.Zero(value.Type())
		}
		item := cloneValue(value.Elem(), seen)
		out := reflect.New(value.Type()).Elem()
		if item.IsValid() {
			out.Set(item)
		}
		return out

	case reflect.Pointer:
		if value.IsNil() {
			return reflect.Zero(value.Type())
		}
		visit := cloneVisit{typ: value.Type(), kind: value.Kind(), ptr: value.Pointer()}
		if prior, ok := seen[visit]; ok {
			return prior
		}
		out := reflect.New(value.Type().Elem())
		seen[visit] = out
		out.Elem().Set(cloneValue(value.Elem(), seen))
		return out

	case reflect.Slice:
		if value.IsNil() {
			return reflect.Zero(value.Type())
		}
		out := reflect.MakeSlice(value.Type(), value.Len(), value.Cap())
		if value.Len() > 0 {
			visit := cloneVisit{typ: value.Type(), kind: value.Kind(), ptr: value.Pointer(), len: value.Len(), cap: value.Cap()}
			if prior, ok := seen[visit]; ok {
				return prior
			}
			seen[visit] = out
		}
		for i := range value.Len() {
			out.Index(i).Set(cloneValue(value.Index(i), seen))
		}
		return out

	case reflect.Map:
		if value.IsNil() {
			return reflect.Zero(value.Type())
		}
		visit := cloneVisit{typ: value.Type(), kind: value.Kind(), ptr: uintptr(value.UnsafePointer())}
		if prior, ok := seen[visit]; ok {
			return prior
		}
		out := reflect.MakeMapWithSize(value.Type(), value.Len())
		seen[visit] = out
		iter := value.MapRange()
		for iter.Next() {
			key := cloneValue(iter.Key(), seen)
			item := cloneValue(iter.Value(), seen)
			out.SetMapIndex(key, item)
		}
		return out

	case reflect.Array:
		out := reflect.New(value.Type()).Elem()
		for i := range value.Len() {
			out.Index(i).Set(cloneValue(value.Index(i), seen))
		}
		return out

	case reflect.Struct:
		// Begin with a value copy so unexported immutable implementation fields
		// (for example time.Time internals) are preserved. Recursively replace
		// every field reflection permits us to set.
		out := reflect.New(value.Type()).Elem()
		out.Set(value)
		for i := range value.NumField() {
			if out.Field(i).CanSet() && value.Field(i).CanInterface() {
				out.Field(i).Set(cloneValue(value.Field(i), seen))
			}
		}
		return out

	default:
		return value
	}
}

func containsSecret(value any) bool {
	return valueContainsSecret(reflect.ValueOf(value), make(map[cloneVisit]struct{}))
}

func valueContainsSecret(value reflect.Value, seen map[cloneVisit]struct{}) bool {
	if !value.IsValid() {
		return false
	}
	if value.Type() == secretType {
		return true
	}

	switch value.Kind() {
	case reflect.Interface:
		return !value.IsNil() && valueContainsSecret(value.Elem(), seen)
	case reflect.Pointer:
		if value.IsNil() {
			return false
		}
		visit := cloneVisit{typ: value.Type(), kind: value.Kind(), ptr: value.Pointer()}
		if _, ok := seen[visit]; ok {
			return false
		}
		seen[visit] = struct{}{}
		return valueContainsSecret(value.Elem(), seen)
	case reflect.Slice:
		if value.IsNil() {
			return false
		}
		if value.Len() > 0 {
			visit := cloneVisit{typ: value.Type(), kind: value.Kind(), ptr: value.Pointer(), len: value.Len(), cap: value.Cap()}
			if _, ok := seen[visit]; ok {
				return false
			}
			seen[visit] = struct{}{}
		}
		for i := range value.Len() {
			if valueContainsSecret(value.Index(i), seen) {
				return true
			}
		}
	case reflect.Array:
		for i := range value.Len() {
			if valueContainsSecret(value.Index(i), seen) {
				return true
			}
		}
	case reflect.Map:
		if value.IsNil() {
			return false
		}
		visit := cloneVisit{typ: value.Type(), kind: value.Kind(), ptr: uintptr(value.UnsafePointer())}
		if _, ok := seen[visit]; ok {
			return false
		}
		seen[visit] = struct{}{}
		iter := value.MapRange()
		for iter.Next() {
			if valueContainsSecret(iter.Key(), seen) || valueContainsSecret(iter.Value(), seen) {
				return true
			}
		}
	case reflect.Struct:
		for i := range value.NumField() {
			if valueContainsSecret(value.Field(i), seen) {
				return true
			}
		}
	}
	return false
}

File: sdk/typescript/src/configstore/clone.ts
import { ReleaseSecret } from "../releases/types.js";
import { Secret } from "../secret.js";

/**
 * Recursively clone supported managed-configuration values while preserving
 * cycles and shared references. Secret plaintext backing storage is never
 * shared with the source.
 */
export function cloneConfig<T>(value: T): T {
  return cloneValue(value, new Map<object, unknown>());
}

/** Return true when a supported value graph contains secret plaintext. */
export function containsSecret(value: unknown): boolean {
  return valueContainsSecret(value, new Set<object>());
}

function cloneValue<T>(value: T, seen: Map<object, unknown>): T {
  if (value === null || (typeof value !== "object" && typeof value !== "function")) {
    return value;
  }
  if (seen.has(value)) return seen.get(value) as T;
  if (value instanceof Secret || value instanceof ReleaseSecret) {
    const result = value.clone();
    seen.set(value, result);
    return result as T;
  }

  if (Buffer.isBuffer(value)) {
    const result = Buffer.from(value);
    seen.set(value, result);
    return result as T;
  }
  if (value instanceof Uint8Array) {
    const result = Uint8Array.from(value);
    seen.set(value, result);
    return result as T;
  }
  if (value instanceof Date) {
    const result = new Date(value.getTime());
    seen.set(value, result);
    return result as T;
  }
  if (value instanceof Map) {
    const result = new Map<unknown, unknown>();
    seen.set(value, result);
    for (const [key, item] of value) {
      result.set(cloneValue(key, seen), cloneValue(item, seen));
    }
    return result as T;
  }
  if (value instanceof Set) {
    const result = new Set<unknown>();
    seen.set(value, result);
    for (const item of value) result.add(cloneValue(item, seen));
    return result as T;
  }
  if (Array.isArray(value)) {
    const lengthDescriptor = Object.getOwnPropertyDescriptor(value, "length");
    if (!lengthDescriptor || !("value" in lengthDescriptor)) {
      throw new TypeError("configstore: array length is not a cloneable data property");
    }
    const result: unknown[] = new Array(lengthDescriptor.value as number);
    seen.set(value, result);
    cloneDataProperties(value, result, seen, true);
    Object.defineProperty(result, "length", lengthDescriptor);
    return result as T;
  }
  if (typeof value === "function") return value;

  const result = Object.create(Object.getPrototypeOf(value)) as object;
  seen.set(value, result);
  cloneDataProperties(value, result, seen, false);
  return result as T;
}

function cloneDataProperties(
  source: object,
  target: object,
  seen: Map<object, unknown>,
  skipArrayLength: boolean,
): void {
  for (const key of Reflect.ownKeys(source)) {
    if (skipArrayLength && key === "length") continue;
    const descriptor = Object.getOwnPropertyDescriptor(source, key);
    if (!descriptor) continue;
    if (!("value" in descriptor)) {
      throw new TypeError("configstore: accessor properties are not cloneable config values");
    }
    Object.defineProperty(target, key, {
      ...descriptor,
      value: cloneValue(descriptor.value, seen),
    });
  }
}

function valueContainsSecret(value: unknown, seen: Set<object>): boolean {
  if (value instanceof Secret || value instanceof ReleaseSecret) return true;
  if (value === null || (typeof value !== "object" && typeof value !== "function")) return false;
  if (seen.has(value)) return false;
  seen.add(value);

  if (value instanceof Map) {
    for (const [key, item] of value) {
      if (valueContainsSecret(key, seen) || valueContainsSecret(item, seen)) return true;
    }
    return false;
  }
  if (value instanceof Set) {
    for (const item of value) {
      if (valueContainsSecret(item, seen)) return true;
    }
    return false;
  }
  for (const key of Reflect.ownKeys(value)) {
    const descriptor = Object.getOwnPropertyDescriptor(value, key);
    if (descriptor && "value" in descriptor && valueContainsSecret(descriptor.value, seen)) {
      return true;
    }
  }
  return false;
}
